### PR TITLE
Display invitations to services in /dashboard

### PR DESCRIPTION
**Pull request by marcel-klasse**
Adds a basic dashboard displaying invitations for each service declared in the config.
Also fetches `dev.medienhaus.meta` for invitations.

- [x] display source (service) of invitation
- [x] display user name who sent invitation

Known Issues:

-  There's a problem with `matrix.invites` showing already handled invitations. To reproduce: Accept invitation for a pad or sketch (click cancel when asked if you want to be redirected) -> reload page.

### CODE REVIEW

File: components/layouts/default.js [hunk 1]
```diff
@@ -10,7 +10,6 @@ export const Main = styled.main`
   @media ${breakpoints.tabletAndAbove} {
     grid-row: 1/4;
     grid-column: 2;
-    max-width: 55ch;
```
Comment (fnwbr) [inline components/layouts/default.js:13]: image
image

File: components/layouts/partials/navigation.js [hunk 2]
```diff
@@ -26,6 +28,8 @@ const List = styled.ul`
 
 export default function Navigation({ closeNavigation }) {
     const auth = useAuth();
+    const matrix = useMatrix(auth.getAuthenticationProvider('matrix'));
```
Comment (fnwbr) [inline components/layouts/partials/navigation.js:31]: Since https://github.com/medienhaus/medienhaus-spaces/pull/44 was merged it's enough to say `useMatrix()` -- there's no parameter that the function accepts anymore 👍

File: components/layouts/partials/navigation.js [hunk 3]
```diff
@@ -26,6 +28,8 @@ const List = styled.ul`
...
+    const invites = matrix.invites;
```
Comment (fnwbr) [inline components/layouts/partials/navigation.js:32]: Is this variable necessary? Or can we just say `matrix.invites.size` down there?

File: pages/chat/[[...roomId]].js [hunk 4]
```diff
@@ -136,7 +136,7 @@ export default function RoomId() {
         };
     });
 
-    const invites = _.sortBy([...matrix.invites.values()], sortRooms);
+    const invites = _.sortBy([...matrix.invites.values()], sortRooms).filter(invite => !invite.meta);
```
Comment (fnwbr) [inline pages/chat/[[...roomId]].js:139]: Can you add a comment what this `.filter(invite => !invite.meta);` call does "in layman terms"?
Is it there to show all invites, _except_ for the ones that we assume to be part of one application?

Which makes me wonder: Wouldn't normal "chats" that we start from the /spaces app also have a `dev.medienhaus.meta` event in the future? Is this filter accurate then?

File: pages/dashboard/DisplayInvitations.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import { ServiceTable } from '../../components/UI/ServiceTable';
+import TextButton from '../../components/UI/TextButton';
+import AcceptIcon from '../../assets/icons/accept.svg';
+import CloseIcon from '../../assets/icons/close.svg';
+import LoadingSpinnerInline from '../../components/UI/LoadingSpinnerInline';
+
+/**
+ * COMPONENT 'DisplayInvitations'
+ *
+ * @TODO
```
Comment (fnwbr) [inline pages/dashboard/DisplayInvitations.js:14]: What's left `@TODO`?

File: pages/dashboard/DisplayInvitations.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+import { useRouter } from 'next/router';
+
+import { ServiceTable } from '../../components/UI/ServiceTable';
+import TextButton from '../../components/UI/TextButton';
+import AcceptIcon from '../../assets/icons/accept.svg';
+import CloseIcon from '../../assets/icons/close.svg';
+import LoadingSpinnerInline from '../../components/UI/LoadingSpinnerInline';
+
+/**
+ * COMPONENT 'DisplayInvitations'
```
Comment (fnwbr) [inline pages/dashboard/DisplayInvitations.js:12]: Might as well remove this comment line... it's not really adding any value, or is it?

File: pages/dashboard/DisplayInvitations.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                { name }
+            </ServiceTable.Cell>
+            <ServiceTable.Cell>
+                { invite.name }
+            </ServiceTable.Cell>
+            <ServiceTable.Cell title={invite.inviter?.userId}>
+                { invite.inviter?.displayName }
+            </ServiceTable.Cell>
+            <>
+                <ServiceTable.Cell title={t('accept invitation')} onClick={(e) => { handleAccept(e, invite.roomId); }}>
```
Comment (fnwbr) [inline pages/dashboard/DisplayInvitations.js:54]: As always: Please add translations for new strings... please always check for this yourselves before asking for a review. Same goes for `decline invitation` below...

File: pages/dashboard/ServiceInvitations.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import _ from 'lodash';
+import getConfig from 'next/config';
+
+import DisplayInvitations from './DisplayInvitations';
+
+/**
+ * COMPONENT 'ApplicationSection'
```
Comment (fnwbr) [inline pages/dashboard/ServiceInvitations.js:7]: Please clean up and add some text for what this component is doing... I think it's worth it here.

File: pages/dashboard/index.js [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+    // const authProviders = getConfig().publicRuntimeConfig.authProviders; // get the authProvider object from the config
+    // const providersWithTemplates = useMemo(() => {
+    //     return _.pickBy(authProviders, provider => _.has(provider, 'templates')); // only return authProviders which have a templates key
+    // }, [authProviders]);
+
+    // functions which interact with matrix server
+    const declineMatrixInvite = async (e, roomId) => {
+        e.preventDefault();
+        console.log('rejecting ' + roomId);
```
Comment (fnwbr) [inline pages/dashboard/index.js:46]: Please use `logger.debug()` instead...

File: pages/dashboard/index.js [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+import { useAuth } from '../../lib/Auth';
+import { useMatrix } from '../../lib/Matrix';
+import ServiceInvitations from './ServiceInvitations';
+import { ServiceTable } from '../../components/UI/ServiceTable';
+import DisplayInvitations from './DisplayInvitations';
+
+const sortRooms = function(room) {
+    return [
+        room.notificationCount === 0,
```
Comment (fnwbr) [inline pages/dashboard/index.js:15]: Doesn't really make sense to sort them by the notification count --- pending invitations always have a `notificationCount` of `1`, I guess. At least it's always the same.

I think it would make more sense to replace the `_.sortBy([...matrix.invites.values()], sortRooms)` line with something like: `_.sortBy([...matrix.invites.values()], 'name')` instead.

File: pages/dashboard/index.js [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+`;
+export default function Dashboard() {
+    const auth = useAuth();
+    const matrix = useMatrix(auth.getAuthenticationProvider('matrix'));
+    const MatrixAuthProvider = auth.getAuthenticationProvider('matrix');
+    const { t } = useTranslation('dashboard');
+
+    const invites = matrix.invites;
+    const matrixClient = auth.getAuthenticationProvider('matrix').getMatrixClient();
+    const serviceSpaces = matrix.serviceSpaces;
```
Comment (fnwbr) [inline pages/dashboard/index.js:35]: Can't see this variable being used...

File: pages/dashboard/index.js [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+  }
+`;
+export default function Dashboard() {
+    const auth = useAuth();
+    const matrix = useMatrix(auth.getAuthenticationProvider('matrix'));
+    const MatrixAuthProvider = auth.getAuthenticationProvider('matrix');
+    const { t } = useTranslation('dashboard');
+
+    const invites = matrix.invites;
+    const matrixClient = auth.getAuthenticationProvider('matrix').getMatrixClient();
```
Comment (fnwbr) [inline pages/dashboard/index.js:34]: I think this is getting a little confusing overall... we have too many layers:

1. `useMatrix()` -- which is referring to things & functions inside `Matrix.js`
2. `auth.getAuthenticationProvider('matrix')` -- referring to functions inside `MatrixAuthProvider.js`
3. `auth.getAuthenticationProvider('matrix').getMatrixClient()` -- referring to the actual "raw" `matrix-js-sdk` `MatrixClient`

Do you have an idea how to simplify this?
One thing I can think of would be to never call (2) from anywhere, except for from `Matrix.js`. Which would mean that all functions we've added in `MatrixAuthProvider.js` (like `addSpaceChild`) would live in `Matrix.js` instead.

File: pages/dashboard/index.js [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+        console.log('rejecting ' + roomId);
+        await matrix.leaveRoom(roomId);
+    };
+
+    const acceptMatrixInvite = async (e, roomId, service, name) => {
+        e.preventDefault();
+        await matrixClient.joinRoom(roomId).catch(() => {
+            return;
+        });
+        await MatrixAuthProvider.addSpaceChild(serviceSpaces[service], roomId);
```
Comment (fnwbr) [inline pages/dashboard/index.js:55]: Sooo... when we accept an invite for e.g. a pad, we add this pad to our "private" Etherpad applications folder? Is that correct? Is this line also being executed for "normal" chats? What does this line do in that case?

File: pages/dashboard/index.js [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    };
+
+    return (
+        <>
+            <h2>/dashboard</h2>
+
+            { /* { invites.size > 0 && <ApplicationSegment>
```
Comment (fnwbr) [inline pages/dashboard/index.js:62]: Remove?

File: pages/dashboard/ServiceInvitations.js [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * @TODO
+ *
+ *
+ * @param {String} name — name of the Application
+ * @param {String} id — id of the application
```
Comment (fnwbr) [inline pages/dashboard/ServiceInvitations.js:13]: This is not accurate... I can see many more parameters down there.
Also, what is the "id of the application"?

File: pages/dashboard/ServiceInvitations.js [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+const ServiceInvitations = ({ service, id, invitations, acceptMatrixInvite, declineMatrixInvite }) => {
+    const name = getConfig().publicRuntimeConfig.authProviders[service].path || service;
+    const serviceTemplates = getConfig().publicRuntimeConfig.authProviders[service].templates;
+    const serviceInvitations = [...invitations.values()].filter(invite => serviceTemplates.includes(invite.meta?.template)); // filter invitations for the current service
+
+    if (_.isEmpty(serviceInvitations)) return null;
+
+    return (
+
```
Comment (fnwbr) [inline pages/dashboard/ServiceInvitations.js:24]: Empty line? Also, no need for the `(` in the line above, or is there?

File: components/UI/ServiceTable.js [hunk 17]
```diff
@@ -42,13 +89,17 @@ const TableCell = styled.td`
 
 `;
 
-export function ServiceTable({ children }) {
+export function ServiceTable({ children, selected, align }) {
```
Comment (fnwbr) [inline components/UI/ServiceTable.js:92]: When are these being used? Can't find any consumers. Shouldn't `ServiceTable` be just this?
```js
const ServiceTable = Table;
```

File: lib/Matrix.js [hunk 18]
```diff
@@ -138,7 +142,20 @@ function useMatrixProvider(auth) {
...
+            const metaEvent = await hydrateMetaEvent(room.roomId);
+            // we check to see who invited the current user
+            const inviterName = room.getMember(matrixClient.getUserId())?.events?.member?.getSender?.();
+            const inviter = matrixClient.getUser(inviterName);
+            setInvites(setRoom(room.roomId, metaEvent, inviter));
+        }
+
+        if (prevMembership === 'invite') {
+            // in this case we know the invitation was handled by the user and we can delete it from our invitations
+            setInvites(deleteElement(room.roomId));
```
Comment (fnwbr) [inline lib/Matrix.js:158]: Isn't line `137` taking care of this already? If not, why not and in which case?

File: lib/Matrix.js [hunk 19]
```diff
@@ -138,7 +142,20 @@ function useMatrixProvider(auth) {
             }
         }
         if ((membership === 'invite')) {
-            setInvites(setRoom(room.roomId));
+            // we need to update the matrixClient room sync so getRooms(roomId) in buildRoomObject doesn't return null
+            await matrixClient.roomInitialSync(room.roomId)
```
Comment (fnwbr) [inline lib/Matrix.js:146]: Find it hard to explain this quickly, but as a rule of thumb: Never do any `async` actions inside the event handler functions.

[These functions](https://github.com/medienhaus/medienhaus-spaces/pull/49/files#diff-2f8d4752f5ecbf336e6d4aeb24e15c2483c854d3d0c24b137625b5ad3376d301L186-L188) get executed for each and every single event of those types that can be found in the Matrix. This includes **all events** during the initial sync at the beginning, and this includes every event of those types that happens afterwards when the user has the page open.

To get around this let's think about another way where we can do this: If I understood correctly you need this in order to display detailed information per invitation on the dashboard.
Do we need that detailed information at all times? Or do we only need it in the dashboard?

If it's the latter, could you think of a way of fetching that `inviter` information only when looking at the dashboard?

If this gets too complicated I'm happy to make these changes myself, after you've addressed the other review comments.

File: next.config.example.js [hunk 20]
```diff
@@ -15,6 +15,7 @@ module.exports = {
                     api: 'http://etherpad.localhost/mypads/api',
                     spacesGroupName: '/spaces', // optional, defaults to publicRuntimeConfig.name
                 },
+                templates: ['etherpad-link'],
```
Comment (fnwbr) [inline next.config.example.js:18]: Is this the correct default? When looking at my rooms from my test instance they all have their template set to `write-link`... I think this is another thing that needs to be addressed with https://github.com/medienhaus/medienhaus-spaces/issues/30 for more consistency.

IMHO this should be just `etherpad`. And udk/spaces knows how to interpret `etherpad`. No matter how the sysadmin/developer decided to call that "app". And `etherpad` will also always be the known by the codebase, as things are called `EtherpadAuthProvider`, `/pages/etherpad/<...>`, and so on here.
This would work **without** defining a template. The template would just always be the exact same as the technical name of the "service".

**Review by fnwbr — CHANGES_REQUESTED** [review]

File: components/layouts/default.js (hunk 1 above)
Comment (marcel-klasse) [reply components/layouts/default.js:13]: i will just @andirueckel here 🙃

File: pages/chat/[[...roomId]].js (hunk 4 above)
Comment (marcel-klasse) [reply pages/chat/[[...roomId]].js:139]: I've added a comment.

> Which makes me wonder: Wouldn't normal "chats" that we start from the /spaces app also have a dev.medienhaus.meta event in the future? Is this filter accurate then?

Yeah this is sort of an ongoing discussion, whether we need a `dev.medienhaus.meta` event for chat rooms or not. I personally don't see the advantage of "spamming" chat rooms with unnecessary events, or to be able to verify the origin? (as in coming from the /spaces app) but I don't really mind either way.

File: components/layouts/default.js (hunk 1 above)
Comment (andirueckel) [reply components/layouts/default.js:13]: The highlighted `max-width`  was removed from the “default layout” on purpose.

We should not, by default, limit the width of our `main` element; this *was* done in another prototype with very different layout and structure. If we want to limit the widths of the sections shown in the above screenshots, then we should do so in their respective components. I’d argue, though, that we might want to adapt the layout logic/structure per se as we made many changes to [medienhaus-spaces](https://github.com/medienhaus/medienhaus-spaces).

File: pages/chat/[[...roomId]].js (hunk 4 above)
Comment (andirueckel) [reply pages/chat/[[...roomId]].js:139]: > Which makes me wonder: Wouldn't normal "chats" that we start from the /spaces app also have a dev.medienhaus.meta event in the future? Is this filter accurate then?

No, we should not add any `dev.medienhaus.meta` events to **chat** rooms.

Chat rooms are considerably well defined, and _none of our business_; we _use_ chat rooms, but we don’t want to re-invent chat rooms or break any compatibility with, well, *any* matrix chat room from any matrix chat client, like for example [element-web](https://github.com/vector-im/element-web).

File: pages/dashboard/DisplayInvitations.js (hunk 5 above)
Comment (marcel-klasse) [reply pages/dashboard/DisplayInvitations.js:14]: updating the JSDoc 😁

File: lib/Matrix.js (hunk 18 above)
Comment (marcel-klasse) [reply lib/Matrix.js:158]: ah yeah, I think this was left over from me trying to fix the weird issue with invitations reappearing.

File: lib/Matrix.js (hunk 19 above)
Comment (marcel-klasse) [reply lib/Matrix.js]: changed it to fetch the information within dashboard.

If we want to display invitations for individual services on the actual service page, as well as on the dashboard it will probably make sense to fetch this inforamtion directly when we receive an invitation. But then let's cross that bridge when we get to it 🙂

File: components/layouts/default.js (hunk 1 above)
Comment (fnwbr) [reply components/layouts/default.js:13]: I'm cool with either way. I'm just not cool with how the `/login` page is looking now. Feel free to implement whatever you prefer, or bring back the `max-width` for now and propose a more general layout change in a separate PR.

**Review by fnwbr — CHANGES_REQUESTED** [review]

File: pages/dashboard/DisplayInvitations.js [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+ * @param e - event triggered by the button
+ * @param roomId - matrix roomId
+ *
+ * @returns {React.ReactComponent}
+*/
+
+const DisplayInvitations = ({ invite, name, service, acceptMatrixInvite, declineMatrixInvite }) => {
+    const { t } = useTranslation('dashboard');
+    const [isAcceptingInvite, setIsAcceptingInvite] = useState(false);
+    const [isDecliningInvite, setisDecliningInvite] = useState(false);
```
Comment (fnwbr) [inline pages/dashboard/DisplayInvitations.js:34]: capital `I` in `setisDecliningInvite` please

File: pages/dashboard/DisplayInvitations.js [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+        setisDecliningInvite(true);
+        await declineMatrixInvite(e, roomId);
+        setisDecliningInvite(false);
+    };
+
+    const handleAccept = async (e, roomId) => {
+        setIsAcceptingInvite(true);
+        await acceptMatrixInvite(e, roomId, service);
+        setIsAcceptingInvite(false);
+        if (confirm('You’ve successfully accepted the invitation!\n\nWould you like to be redirected to the newly accepted ' + name + ' item?')) {
```
Comment (fnwbr) [inline pages/dashboard/DisplayInvitations.js:47]: Missing translations...

File: pages/dashboard/DisplayInvitations.js [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+
+const DisplayInvitations = ({ invite, name, service, acceptMatrixInvite, declineMatrixInvite }) => {
+    const { t } = useTranslation('dashboard');
+    const [isAcceptingInvite, setIsAcceptingInvite] = useState(false);
+    const [isDecliningInvite, setisDecliningInvite] = useState(false);
+    const router = useRouter();
+
+    const handleDecline = async (e, roomId) => {
+        setisDecliningInvite(true);
+        await declineMatrixInvite(e, roomId);
```
Comment (fnwbr) [inline pages/dashboard/DisplayInvitations.js:39]: Instead of forwarding the `e` only to prevent the default click handler logic, please do that here.
As a rule of thumb: Catch those default behaviors always as close as possible to the original sending element. In this case it's the table cell. And not something that's defined in the file where you're currently calling `e.preventDefault()`. This will also allow you to not having to pass the `e` down the call stack all the way.

File: pages/dashboard/DisplayInvitations.js [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+        setisDecliningInvite(false);
+    };
+
+    const handleAccept = async (e, roomId) => {
+        setIsAcceptingInvite(true);
+        await acceptMatrixInvite(e, roomId, service);
+        setIsAcceptingInvite(false);
+        if (confirm('You’ve successfully accepted the invitation!\n\nWould you like to be redirected to the newly accepted ' + name + ' item?')) {
+            router.push(`${name}/${roomId}`);
```
Comment (fnwbr) [inline pages/dashboard/DisplayInvitations.js:48]: Instead of _assuming_ that this path pattern always works, can we make the corresponding consumer define it? Something I can think of, for example, would be to return it in the `acceptMatrixInvite` function like this:
```js
const forwardingUrl = await acceptMatrixInvite(e, roomId, service);

...

if (...) { router.push(forwardingUrl); }
```

File: pages/dashboard/DisplayInvitations.js (hunk 23 above)
Comment (fnwbr) [reply pages/dashboard/DisplayInvitations.js:39]: (Same goes for `acceptMatrixInvite` below)

**Review by fnwbr — CHANGES_REQUESTED** [review]

File: pages/chat/[[...roomId]].js (hunk 4 above)
Comment (fnwbr) [reply pages/chat/[[...roomId]].js:147]: Cool, agreed then. 👍  Let's just add a short comment on what the filtering is supposed to do. Feel free to leave a link to _this_ very discussion here on GitHub inside the code.

File: pages/dashboard/DisplayInvitations.js [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+            <ServiceTable.Cell title={invite.inviter?.userId}>
+                { invite.inviter?.displayName }
+            </ServiceTable.Cell>
+            <>
+                <ServiceTable.Cell title={t('accept invitation')} onClick={(e) => { handleAccept(e, invite.roomId); }}>
+                    <TextButton disabled={isDecliningInvite || isAcceptingInvite}>
+                        { isAcceptingInvite ? <LoadingSpinnerInline /> : <AcceptIcon /> }
+                    </TextButton>
+                </ServiceTable.Cell>
+                <ServiceTable.Cell title={t('decline invitation')} onClick={(e) => {handleDecline(e, invite.roomId);}}>
```
Comment (fnwbr) [inline pages/dashboard/DisplayInvitations.js:69]: A quick detour into accessibility land: As a rule of thumb you don't want to add `onClick` handlers to anything but a `<button>` (or `<a>` elements, when changing the URL)...
The short solution would be to add this to the `<TextButton>` element below instead.

If you want to read up on this:
- https://github.com/reactjs/react-a11y/blob/master/docs/rules/onclick-uses-role.md (and the W3 article linked at the bottom)
- https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#accessibility_concerns

**Review by fnwbr — CHANGES_REQUESTED** [review]

File: pages/dashboard/index.js (hunk 12 above)
Comment (marcel-klasse) [reply pages/dashboard/index.js:28]: maybe i'll change this in a separate PR? the same would go for `removeSpaceChild`

File: pages/dashboard/DisplayInvitations.js [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+        setIsDecliningInvite(false);
+    };
+
+    const handleAccept = async (e, roomId) => {
+        e.preventDefault();
+        setIsAcceptingInvite(true);
+        const forwardingUrl = await acceptMatrixInvite(roomId, service);
+        setIsAcceptingInvite(false);
+        if (forwardingUrl) {
+            if (confirm('You’ve successfully accepted the invitation!\n\nWould you like to be redirected to the newly accepted {{name}} item?', { name: path })) {
```
Comment (fnwbr) [inline pages/dashboard/DisplayInvitations.js:48]: I'm not sure if I like this... let's assume I have 10 invitations. So instead of being able to quickly accept/reject them, I have to confirm or deny this after every step.

It's more like this: either I want to be forwarded to the item I just accepted the invitation for, or I don't and I want to stay on my dashboard. But if I choose the latter, why would I decide otherwise after accepting/rejecting the next invitation?

I'm not sure what to suggest but removing this `confirm()` altogether.
Or somehow memorizing if the user said "no, don't forward me" and then not doing so for all subsequent accepts.

What do you think from an UX POV @andirueckel?

Comment (fnwbr) [reply pages/dashboard/DisplayInvitations.js:48]: Also, something's not working:
image

**Review by fnwbr — CHANGES_REQUESTED** [review]
1. After accepting a `/chat` invitation I got redirected to this:
  image
2. Why does this say `write`? Shouldn't there be a `/` in front? I'm using the same config as `next.config.example.js`.
  image

File: pages/dashboard/DisplayInvitations.js (hunk 26 above)
Comment (andirueckel) [reply pages/dashboard/DisplayInvitations.js:48]: > It's more like this: either I want to be forwarded to the item I just accepted the invitation for, or I don't and I want to stay on my dashboard. But if I choose the latter, why would I decide otherwise after accepting/rejecting the next invitation?

@fnwbr @marcel-klasse Let’s replace the `confirm()` with `alert()` and remove the ”Would you like to be redirected …” paragraph. After having accepted the invitation for/to an item, let’s disable the respective accept/decline buttons and make the item’s display name a proper hyperlink.

File: pages/dashboard/DisplayInvitations.js [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+            <ServiceTable.Cell>
+                { path }
+            </ServiceTable.Cell>
+            <ServiceTable.Cell>
+                { invite.name }
+            </ServiceTable.Cell>
+            <ServiceTable.Cell title={invite.inviter?.userId}>
+                { invite.inviter?.displayName }
+            </ServiceTable.Cell>
+            <>
```
Comment (fnwbr) [inline pages/dashboard/DisplayInvitations.js:61]: Why is this empty element here?

File: pages/dashboard/DisplayInvitations.js [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+                </ServiceTable.Cell>
+                <ServiceTable.Cell title={t('decline invitation')}>
+                    <TextButton onClick={(e) => {handleDecline(e, invite.roomId);}} disabled={isDecliningInvite || isAcceptingInvite}>
+                        { isDecliningInvite ? <LoadingSpinnerInline /> : <CloseIcon /> }
+                    </TextButton>
+                </ServiceTable.Cell>
+            </>
+
+        </ServiceTable.Row>
+
```
Comment (fnwbr) [inline pages/dashboard/DisplayInvitations.js:75]: Empty line

File: pages/dashboard/DisplayInvitations.js [hunk 29]
```diff
@@ -0,0 +1,78 @@
...
+                        { isAcceptingInvite ? <LoadingSpinnerInline /> : <AcceptIcon /> }
+                    </TextButton>
+                </ServiceTable.Cell>
+                <ServiceTable.Cell title={t('decline invitation')}>
+                    <TextButton onClick={(e) => {handleDecline(e, invite.roomId);}} disabled={isDecliningInvite || isAcceptingInvite}>
+                        { isDecliningInvite ? <LoadingSpinnerInline /> : <CloseIcon /> }
+                    </TextButton>
+                </ServiceTable.Cell>
+            </>
+
```
Comment (fnwbr) [inline pages/dashboard/DisplayInvitations.js:73]: Empty line

File: pages/dashboard/ServiceInvitations.js [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+    return _.map(serviceInvitations, (invite) => {
+        return <DisplayInvitations
+            key={invite.roomId}
+            service={service}
+            path={path}
+            invite={invite}
+            declineMatrixInvite={declineMatrixInvite}
+            acceptMatrixInvite={acceptMatrixInvite} />;
+    })
+
```
Comment (fnwbr) [inline pages/dashboard/ServiceInvitations.js:34]: Empty line... and the line below can go to line 33

**Review by fnwbr — CHANGES_REQUESTED** [review]

File: public/locales/de/_defaults.json [hunk 31]
```diff
@@ -8,5 +8,9 @@
   "Save changes": "Änderungen speichern",
   "Select action": "Aktion wählen",
   "What would you like to do?": "Was möchtest Du tun?",
-  "{{error}}, please wait {{time}} seconds.": "{{error}}, bitte warte {{time}} Sekunden."
-}
\ No newline at end of file
+  "{{error}}, please wait {{time}} seconds.": "{{error}}, bitte warte {{time}} Sekunden.",
+  "Invitations": "Einladungen",
+  "From": "Von",
+  "Accept": "Annehemen",
```
Comment (fnwbr) [inline public/locales/de/_defaults.json:14]: image

**Review by fnwbr — CHANGES_REQUESTED** [review]

**Review by fnwbr — APPROVED** [review]
